### ISSUE TITLE
XEP-0425: Message Moderation (in MUC)

### ISSUE THREAD

**Issue by Zash** [dino/dino#1133]
Sometimes people post inappropriate content. XMPP has historically not had much more than IRC in terms of moderation tooling, ie kicking and banning, but today that might be considered insufficient.

[XEP-0425: Message Moderation](https://xmpp.org/extensions/xep-0425.htm) describes a protocol for a moderator to moderate individual messages, which would be nice to support.

Implementing enough of [XEP-0425](https://xmpp.org/extensions/xep-0425.html#usecase-success) to reflect when messages get removed by moderator might be a sensible first step. This allows moderation to be performed, either with specialized tools or other clients, as well as gathering experience which could feed back into getting the XEP moving forward.

**Comment by diodelass** [comment]
It would be fantastic to see a client offer good UI support for moderation tools. At the moment, I don't believe I know of any that do, which makes public MUCs effectively nonviable. I have even been present for a serious incident that happened in a small private MUC where such drama wasn't expected to happen, and the fact that none of us present had a client supporting moderation functions seriously hampered our ability to respond. It was the sort of incident that would have been over in seconds on an IRC channel, but current XMPP clients just don't seem to have the necessary functions (or at least have them deeply buried).

Dino could do a lot of good by adding an easy way to use these functions.

**Comment by tuxayo** [comment]
>  I have even been present for a serious incident that happened in a small private MUC where such drama wasn't expected to happen, and the fact that none of us present had a client supporting moderation functions seriously hampered our ability to respond.

Yes not having XEP-0425 (does it mandates deletion of downloaded images?) can allow terrible scenarios:

That's a way to place any imaginable of the most illegal materials on many people's devices. On Conversations, images by default are even visible in the image gallery. It can even be used to target people with legal harassment, sure it's likely to not end up with a conviction, but that's still trouble with the authorities and likely mistreatment by them by default. And can lead to illegal but legitimate stuff to be compromised due the likely search in the victim's devices. If I were a gov surveillance agency I could even use the whole trick to get an opportunity to bug (hardware or software) someone's devices and/or home during the search.
Another horror scenario is revenge porn, the victim standing here powerless knowing that even if ban was fast, the images have been sent to many people's devices and will still be available for download for those with autodownload off.

Public chats with images open such a big can of worms.

**Comment by pravi** [comment]
For now Gajim, Monocles Chat and Cheogram already support it.

Would anyone be interested to implement this feature if we are able to raise some funds?

**Comment by vpzomtrrfrt** [comment]
Looks like there are multiple different versions of this XEP, and Prosody's mod_muc_moderation doesn't implement the latest one. The previous version depends on XEP-0422, and I'm not sure what might break if we start advertising support for that

**Comment by vpzomtrrfrt** [comment]
actually nevermind, I was looking at an old version of mod_muc_moderation, supposedly the latest one _does_ support 0.3.0

**Comment by fiaxh** [comment]
An initial implementation of message moderation is done in 1ede8ad.

**Comment by pravi** [comment]
Thanks, this was the big missing feature - especially during spam waves. I hope we can see a new release soon.
